### PR TITLE
Ability to save multi-animal pose tracks to single-animal files

### PR DESCRIPTION
**Pull request by DhruvSkyy**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**What does this PR do?**

This PR adds the functionality to save xarrays as a dictionary of individual pandas dataframes for to_dlc_df and as multiple files for each individual for to_dlc_file. This is based on the split_individuals: bool parameter, if True it will split the data for each individual, if false it will save it as a multi-animal file. 

This issue was described in Issue #39. 

## References

Please reference any existing issues/PRs that relate to this PR.

## How has this PR been tested?

The PR was tested on a Juypter Notebook against the four scenarios, (credit to @niksirbi):
* The xarray.Dataset has multiple (>1) individuals and the user wants to save it to a multi-animal DeepLabCut dataframe. This currently works with the existing functions.
* The xarray.Dataset has only 1 individual but the user still wants to save it to a multi-animal DeepLabCut dataframe, including the redundant "individuals" level. This also works currently.
* The xarray.Dataset has a single individual and the user wants to save it to single-animal DeepLabCut dataframe (without the "individuals" level). This behaviour needs to be implemented, the output will be a single file.
* The xarray.Dataset has multiple (>1) individuals and the user wants to split them for saving into multiple single-animal files. * This is the trickiest case to implement, as the dataset would have to be split across individuals (behind the scenes) and each part is then saved to a single-animal file as in case 3. The output should be as many files as there are individuals.

## Todo

Need to write tests. 

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: movement/io/save_poses.py [hunk 1]
```diff
@@ -104,9 +186,50 @@ def to_dlc_file(ds: xr.Dataset, file_path: Union[str, Path]) -> None:
...
+    if split_individuals:
+        dfdict = to_dlc_df(ds, True)
+        if file.path.suffix == ".csv":
+            for (
+                key,
+                df,
+            ) in dfdict.items():
+                """Iterates over dictionary, the key is the name of the
+                individual and the value is the corresponding df."""
+                filepath = (
```
Comment (niksirbi) [inline movement/io/save_poses.py:208]: I find f-strings more readable, so I would rewrite this as:
```python
filepath = f"{file.path.with_suffix('')}_{key}.csv"
df.to_csv(Path(filepath), sep=",")
```

File: movement/io/save_poses.py [hunk 2]
```diff
@@ -104,9 +186,50 @@ def to_dlc_file(ds: xr.Dataset, file_path: Union[str, Path]) -> None:
...
+    if split_individuals is None:
+        individuals = ds.coords["individuals"].data.tolist()
+        if len(individuals) == 1:
+            split_individuals = True
+        else:
+            split_individuals = False
+
+    """If split_individuals is True then it will split the file into a
+    dictionary of pandas dataframes for each individual."""
+    if split_individuals:
```
Comment (niksirbi) [inline movement/io/save_poses.py:199]: Again here, as in the above function, we have to check if there are more than one individuals to split, otherwise output only one single-animal file.

File: movement/io/save_poses.py [hunk 3]
```diff
@@ -104,9 +186,50 @@ def to_dlc_file(ds: xr.Dataset, file_path: Union[str, Path]) -> None:
...
-    df = to_dlc_df(ds)  # convert to pandas DataFrame
-    if file.path.suffix == ".csv":
-        df.to_csv(file.path, sep=",")
-    else:  # file.path.suffix == ".h5"
-        df.to_hdf(file.path, key="df_with_missing")
-    logger.info(f"Saved PoseTracks dataset to {file.path}.")
+    # Sets default behaviour for the function
+    if split_individuals is None:
+        individuals = ds.coords["individuals"].data.tolist()
+        if len(individuals) == 1:
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/io/save_poses.py:192]: We also may want to throw an error if the user passes an invalid type, something like:
```python
if split_individuals == "auto":
        individuals = ds.coords["individuals"].data.tolist()
        split_individuals = True if len(individuals) > 1 else False
    elif not isinstance(split_individuals, bool):
        error_msg = (
            f"Expected 'split_individuals' to be a boolean or 'auto', but got "
            f"{type(split_individuals)}."
        )
        log_error(ValueError, error_msg)
```

**Review by niksirbi — CHANGES_REQUESTED** [review]
Hey @DhruvSkyy, thanks a lot for contributing!
I left several comments. The main two points to be addressed are:
1. The two functions are now quite big and hard to parse, I think we would benefit from refactoring repeated bits into separate functions. I have pointed this out in some concrete examples.
4. We don't really handle the case where there is one individual in the dataset but the user has chosen `split_individuals=True`. I think in this case we should save one single-animal DataFrame (without the "individuals" level), as there is nothing to split. But, we should raise a warning to tell the user that we are ignoring the `split_individuals` setting. 

Let me know if you think these are sensible choices. Have a shot at implementing them and we can have another round of review then.

File: movement/io/save_poses.py [hunk 4]
```diff
@@ -11,72 +11,133 @@
...
-        index=np.arange(ds.dims["time"], dtype=int),
-        columns=columns,
-        dtype=float,
-    )
-    logger.info("Converted PoseTracks dataset to DLC-style DataFrame.")
-    return df
-
-
-def to_dlc_file(ds: xr.Dataset, file_path: Union[str, Path]) -> None:
+    if split_individuals:
```
[the remark this answers is not in the document]
Comment (DhruvSkyy) [reply movement/io/save_poses.py:63]: Hi @niksirbi, 

If split_individuals == True and we have only one individual, it will just output one single-animal DataFrame (with "scorer", "bodyparts", "coords"). If split_individuals == False and we have only one individual, it will output one single-animal DataFrame (with "scorer", "individuals", "bodyparts", "coords"). 

Although for the second case when split_individuals == False the individuals column will just be filled with one individual, it might still be important to have this feature. It might be useful in case they wanted to merge the dataframes with other multi-individual dataframes, where pandas would want the dataframes to have the same format. 

It also might be unexpected if they ran the function with split_individuals == False on a set of data with a mixture of single and multi-individual xarrays and saw the output being a mixture of single-animal and multi-animal dataframes as the single-individual xarrays would automatically turn into single-individual dataframes with no choice to make it multi-individual. 

The auto function in to_dlc_file handles cases when the user might want all single individual xarrays to be stored as single-individual dataframes and multi-individual xarrays to be stored as multi-individual dataframes, I can make a separate function for this auto feature and also use it for to_dlc_df if preferable.

File: movement/io/save_poses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DhruvSkyy) [reply movement/io/save_poses.py:192]: For the auto function, would we want it to save a single individual xarray as a single individual dataframe and a multi-individual xarray as a multi-individual dataframe, or save both as single individual dataframes?

File: movement/io/save_poses.py (hunk 2 above)
Comment (DhruvSkyy) [reply movement/io/save_poses.py:199]: If split_individuals is True and there is only one individual to split it should already automatically output only one single-animal file.

File: movement/io/save_poses.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/io/save_poses.py:63]: > If split_individuals == True and we have only one individual, it will just output one single-animal DataFrame (with "scorer", "bodyparts", "coords"). If split_individuals == False and we have only one individual, it will output one single-animal DataFrame (with "scorer", "individuals", "bodyparts", "coords").

Hm, I actually like this suggestion and the flexibility in gives to the user. The way you are proposing means that `split_individuals == True` will always give an output `csv` in "single-animal" format, while `split_individuals == False` will always return a csv in "multi-animal" format, regardless of how many animals are in the project.

The arguments you make for it are convincing, so let's go ahead and do this! We just have to be careful to write the docstrings in an understandable way.
